### ISSUE TITLE
new xdsl_driver Fritzbox 7582 V7.12

### ISSUE THREAD

**Issue by tjeh** [Ansuel/tch-nginx-gui#785]
Hi,

which files are required to integrate the adsl driver from the new firmware [Fritzbox 7582 Version 7.12](http://download.avm.de/fritzbox/fritzbox-7582/other/fritz.os/FRITZ.Box_7582-07.12.image)?

Is it sufficient to just replace DGA4132 `/etc/adsl/adsl_phy.bin` with the Fritzbox `adsl_phy_anxa.bin` as in `cp /tmp/adsl_phy_anxa.bin /etc/adsl/adsl_phy.bin` ?

**Comment by FrancYescO** [comment]
Basically yes, there are a few other versions (the annexB and "old" variant) integrated in the fritz firmware as diff file.. if you want make a PR to add these

**Comment by tjeh** [comment]
I am not familiar with a PR of such kind, yet. 
However here are the files from the firmware image:  [`/etc/adsl`](https://github.com/Ansuel/tch-nginx-gui/files/3538561/adsl.zip)

- `adsl_phy_anxa.bin`
- `adsl_phy_anxa_ch.bin.bsdiff`
- `adsl_phy_anxa_ch.bin.md5sum`
- `adsl_phy_anxb.bin.bsdiff`
- `adsl_phy_anxb.bin.md5sum`

Are the [`/lib/modules/4.1.38/kernel/drivers/brcm_adsl_drv/`](https://github.com/Ansuel/tch-nginx-gui/files/3538573/brcm_adsl_drv.zip) required, too?

- `adsldd_anxa.ko`
- `adsldd_anxa_ch.ko`
- `adsldd_anxb.ko`

If you give me a hint on how to properly create the PR I am happy to put in the effort.

**Comment by FrancYescO** [comment]
For now just ignore .ko (these are needed but in my last try to load different kernel modules they just refused to load)

drivers binary must but putted here https://github.com/Ansuel/tch-nginx-gui/tree/master/xdsl_driver/arm

Than place md5 of binary drivers files here
https://github.com/Ansuel/tch-nginx-gui/blob/master/decompressed/gui_file/usr/share/transformer/scripts/refresh_driver.sh

Finally edit the modal
https://github.com/Ansuel/tch-nginx-gui/blob/master/decompressed/gui_file/www/docroot/modals/xdsl-low-modal.lp#L194

PS. To get version of the driver without placing it on the modem you could use `strings adsl_phy.bin | grep on`
PS2. `bspatch` must be used to apply the .bsdiff file to the main .bin file, if i'm not wrong in my last try I've used this python alternative https://github.com/graingert/python-bsdiff

**Comment by tjeh** [comment]
ok so `adsl_phy_anxa.bin` is:
`Broadcom DSL Version A2pvfbH045l`
`Built on 03/15/2019 12:03:13`

Ok I have completed the steps for the file above. Shall I also add the following files? 

`adsl_phy_anxa_ch.bin`
`IBroadcom DSL Version A2pvfbH043j2`
`Built on 03/12/2018 12:12:39`

`adsl_phy_anxb.bin.bsdiff`
`Broadcom DSL Version B2pvfbH045l`
`Built on 03/15/2019 11:56:37`

**Comment by FrancYescO** [comment]
Looks like 43j2 already in repo, so just add both 45l

Ah, if you want to try to load the kernel module (now that we have a 4.1 kernel we can have more possibilities load will go okay)
`rmmod adsldd`
`insmod /tmp/adsldd.ko` (the one from fritz)
Than restart adslctl

Replacing directly the module in /lib/modules/ can cause bootloops if the module cause a kernel crash, so loading it from /tmp (also a symlink from /lib/modules can be made) should be safe

**Comment by tjeh** [comment]
so to be clear:

1. put adsldd on my DGA4132
2. `rmmod adsldd`
3. `insmod /tmp/adsldd.ko` 

and report back? Worst case I need to reboot the DGA4132?

**Comment by FrancYescO** [comment]
Yep (i'm not sure insmod accept filepath, if not you should follow the symlink trick)

**Comment by tjeh** [comment]
ok... so this went south pretty fast. loading the `adsldd_anxa.ko` crashed the router.

**Comment by FrancYescO** [comment]
why you had to reflash? have you putted the .ko in /tmp ? is also strange that a reflash (tftp) helped as tftp does not delete the overlay

what have you noticed about ipv6? we have few issues ongoing about ipv6 like https://github.com/Ansuel/tch-nginx-gui/issues/749 that needs to be addressed (and probably a correct rebase with 2.2.0 can help)

**Comment by tjeh** [comment]
So there should probably be a new topic for the 2.2.0 rebase.

As to the `adsldd_anxa.ko`:
1. copied to /tmp
2. symlinked into modules
3. `rmmod adsldd`
4. `insmod adsldd_anxa.ko`
5. reboot
Result: GUI not accessible, after Reset (Pin at the back of the device) GUI 1.0.3 in Bank_2 showed up. Therefore I activated Bank_1, applied GUI.tar_gz and applied 2.2.0 firmware update through Ansuel GUI.

**Comment by FrancYescO** [comment]
this can be a naming issue, another try is to load it named adsldd (the symlink should replace the original adsldd.ko in /lib/modules) so also issuing `insmod adsldd` should work to load the module, than if the module (from fritz) is in /tmp once the modem is rebooted it should just fail to load the module as the binary in /tmp will disappear once rebooted but should continue to work (without xdsl until the original adsldd.ko is placed back and loaded)

**Comment by tjeh** [comment]
Will give it another try later. What do you expect from the updated adsldd. I suspect that the updated file will require update of its dependend modules as well?

**Comment by FrancYescO** [comment]
the kernel module and driver are build to work together, while just hard-replacing only the driver work the kernel module can "search for function" that the replaced driver does not have causing instability or others unwanted behaviours
